transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,486 @@ fn test_input_rules_verify_max_tombstone() {
+
+    // Add a partial fill output, by doubling one of the revealed tx outs.
+    get_first_rules_mut(&mut tx)
+        .partial_fill_outputs
+        .push(change_committed_amount(
+            &revealed_tx_outs[1],
+            Amount::new(2000, 0.into()),
+        ));
+    // Add a parital fill change, also by doubling one of the revealed tx outs.
+    // This means the fill fraction is 1/2, which we are satisfying.
+    get_first_rules_mut(&mut tx).partial_fill_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(2000, 0.into()),
+    ));
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Change the partial fill change value so that the amount shared secret doesn't
+    // match what is recorded
+    let amount_shared_secret = [9u8; 32];
+    get_first_rules_mut(&mut tx)
+        .partial_fill_change
+        .as_mut()
+        .unwrap()
+        .tx_out
+        .masked_amount = Some(
+        MaskedAmount::new_from_amount_shared_secret(
+            block_version,
+            Amount::new(2000, 0.into()),
+            &amount_shared_secret,
+        )
+        .unwrap(),
+    );
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RevealedTxOut(RevealedTxOutError::Amount(_)))
+    );
+
+    // Now the amount shared secret matches what is recorded, but it doesn't match
+    // the real change output
+    get_first_rules_mut(&mut tx)
+        .partial_fill_change
+        .as_mut()
+        .unwrap()
+        .amount_shared_secret = amount_shared_secret.to_vec();
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RevealedTxOut(RevealedTxOutError::Amount(_)))
+    );
+
+    // Now the amount shared secret is the wrong size
+    get_first_rules_mut(&mut tx)
+        .partial_fill_change
+        .as_mut()
+        .unwrap()
+        .amount_shared_secret = Default::default();
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RevealedTxOut(
+            RevealedTxOutError::InvalidAmountSharedSecret
+        ))
+    );
+}
+
+// Test that input rules verification is working for a Tx with partial fill
+// outputs and various fill fractions
+#[test]
+fn test_input_rules_verify_partial_fill_outputs() {
+    let block_version = BlockVersion::THREE;
+    let (mut tx, revealed_tx_outs) = get_input_rules_test_tx(block_version);
+
+    // Add a partial fill output, by tripling one of the revealed tx outs.
+    // So this means we are filling the 1/3 of the request.
+    get_first_rules_mut(&mut tx)
+        .partial_fill_outputs
+        .push(change_committed_amount(
+            &revealed_tx_outs[1],
+            Amount::new(3000, 0.into()),
+        ));
+
+    // Set the partial fill change output to be 1500. This means the implied fill
+    // fraction is now 1/3. This should be valid.
```

Comment:
we're not giving 3000, what's happening here is:

* All actual TxOut's in the transaction have a value of 1000
* We've now specified a partial_fill_change output by taking one of those and making it's amount be 1500

When validation runs, it will see the partial fill change rule for 1500, and find a corresponding output for 1000. So we gave back 2/3, and we took 1/3 of the partial_fill_change. So the fill fraction is 1/3

Reply:
For the sake of convenience I'm going to refer to the stuff Originator wants to receive as a result of this transaction as token2, and the stuff that the originator is putting up as token1 even though I think in this case they're all token0?

By setting partial fill amount to 3000, we are saying that the originator wants 3000 of token2.
By setting partial fill change to 1500, we are saying that the originator is providing up to 1500 of token1. Thus an exchange rate of 2 to 1.
By default of get_input_rules_test_tx the originator is receiving 1000 of token 2 as output (1/3 of their request), and 1000 of token1 as change (2/3 of their original stake)

This should verify, because 1/3 of their stake is being spent, and 1/3 of their request is being given to them.

If we set their requested output to 3001, then we are giving them less than 1/3 of their request, but spending 1/3 of their stake, so that should reject.

If we set the partial fill change to 1501, then because we are sending back 1000, we are effectively taking 501 of token1, and only giving back 1000/3000 of token2.

if the originator wants 2997 token2, and is putting up a stake of 1501 token 1, then giving them back 1000 token1 means we are taking 501/1501 token1. Which means we need to be giving back slightly more than 1000 token2 in order to make this even. So only giving 1000 should reject.

if the originator wants 2996 token2, and is putting up a stake of 1501 token 1, then giving them back 1000 token1 means we are taking 501/1501 token1. Which means we need to be giving back slightly more than 999.99 token2 in order to make this even. So only giving 1000 should accept..

Reply:
I think rubber ducking here has explained it to me. I was just confusing myself.